File: backend/src/HIS.Core/Entities/BillingGuarantor.cs
namespace HIS.Core.Entities;

/// <summary>
/// Đơn vị bảo lãnh viện phí (công ty, tổ chức thanh toán thay bệnh nhân).
/// </summary>
public class SponsorOrg : BaseEntity
{
    /// <summary>Mã đơn vị bảo lãnh (duy nhất trong hệ thống).</summary>
    public string Code { get; set; } = string.Empty;

    /// <summary>Tên đơn vị bảo lãnh.</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>Mã số thuế.</summary>
    public string? TaxCode { get; set; }

    /// <summary>Địa chỉ.</summary>
    public string? Address { get; set; }

    /// <summary>Người liên hệ.</summary>
    public string? ContactPerson { get; set; }

    /// <summary>Số điện thoại liên hệ.</summary>
    public string? Phone { get; set; }

    /// <summary>Còn hoạt động hay không.</summary>
    public bool IsActive { get; set; } = true;
}

/// <summary>
/// Gán bảo lãnh viện phí cho một lượt khám/nội trú.
/// Không dùng FK navigation để tránh cascade phức tạp — chỉ lưu Guid.
/// </summary>
public class BillingGuarantor : BaseEntity
{
    /// <summary>Id đơn vị bảo lãnh (SponsorOrg.Id — không FK).</summary>
    public Guid SponsorOrgId { get; set; }

    /// <summary>Id bệnh nhân (tùy chọn — không FK).</summary>
    public Guid? PatientId { get; set; }

    /// <summary>Id hồ sơ bệnh án / lượt nội trú (tùy chọn — không FK).</summary>
    public Guid? MedicalRecordId { get; set; }

    /// <summary>Số công văn / giấy bảo lãnh.</summary>
    public string? GuaranteeNo { get; set; }

    /// <summary>Tỷ lệ bảo lãnh (0-100%).</summary>
    public decimal GuaranteeRate { get; set; }

    /// <summary>Hạn mức tiền bảo lãnh (null = không giới hạn).</summary>
    public decimal? GuaranteeAmount { get; set; }

    /// <summary>Ngày bắt đầu hiệu lực.</summary>
    public DateTime? FromDate { get; set; }

    /// <summary>Ngày hết hiệu lực.</summary>
    public DateTime? ToDate { get; set; }

    /// <summary>Trạng thái: 0=Hiệu lực, 1=Hết hạn, 2=Hủy.</summary>
    public int Status { get; set; } = 0;

    /// <summary>Ghi chú.</summary>
    public string? Notes { get; set; }
}

File: frontend/src/api/billingGuarantor.ts
import apiClient from './client';

const BASE = '/billing-guarantor';

// ─── DTO interfaces ──────────────────────────────────────────────────────────

export interface SponsorOrgDto {
  id: string;
  code: string;
  name: string;
  taxCode?: string;
  address?: string;
  contactPerson?: string;
  phone?: string;
  isActive: boolean;
}

export interface BillingGuarantorDto {
  id: string;
  sponsorOrgId: string;
  /** join từ SponsorOrg — chỉ đọc */
  sponsorOrgName?: string;
  patientId?: string;
  medicalRecordId?: string;
  guaranteeNo?: string;
  /** Tỷ lệ bảo lãnh 0-100 (%) */
  guaranteeRate: number;
  /** Hạn mức tiền (null = không giới hạn) */
  guaranteeAmount?: number;
  fromDate?: string;
  toDate?: string;
  /** 0=Hiệu lực, 1=Hết hạn, 2=Hủy */
  status: number;
  notes?: string;
}

export interface GuarantorDebtReportDto {
  sponsorOrgId: string;
  sponsorOrgName: string;
  totalGuaranteed: number;
  count: number;
}

// ─── SponsorOrg ─────────────────────────────────────────────────────────────

export const getSponsorOrgs = (keyword?: string) =>
  apiClient.get<SponsorOrgDto[]>(`${BASE}/sponsor-orgs`, { params: { keyword } }).then(r => r.data);

export const saveSponsorOrg = (dto: Partial<SponsorOrgDto>) =>
  apiClient.post<SponsorOrgDto>(`${BASE}/sponsor-orgs`, dto).then(r => r.data);

export const deleteSponsorOrg = (id: string) =>
  apiClient.delete(`${BASE}/sponsor-orgs/${id}`).then(r => r.data);

// ─── BillingGuarantor ───────────────────────────────────────────────────────

export const getGuarantors = (params?: {
  patientId?: string;
  medicalRecordId?: string;
  keyword?: string;
}) =>
  apiClient.get<BillingGuarantorDto[]>(`${BASE}/guarantors`, { params }).then(r => r.data);

export const saveGuarantor = (dto: Partial<BillingGuarantorDto>) =>
  apiClient.post<BillingGuarantorDto>(`${BASE}/guarantors`, dto).then(r => r.data);

export const deleteGuarantor = (id: string) =>
  apiClient.delete(`${BASE}/guarantors/${id}`).then(r => r.data);

// ─── Debt report ─────────────────────────────────────────────────────────────

export const getDebtReport = (from?: string, to?: string) =>
  apiClient.get<GuarantorDebtReportDto[]>(`${BASE}/debt-report`, { params: { from, to } }).then(r => r.data);
